documentation/role-requirements.md

```diff
@@ -0,0 +1,141 @@
+---
+layout: page
+title: "How to Specify Roles Used by Other Roles"
+---
+# Intro
+
+Many of our roles need to use the functionality provided by another role.
+
+An example for the Logging role:
+* When new certs need to be generated with deploying the logging system,
+  integration with the certificate role. See also
+  (Certificate README)[https://github.com/linux-system-roles/certificate/blob/master/README.md]
+* A RELP connection may need to open a port in the firewall, and could use the
+  firewall role.
+* A RELP connection may need to set SELinux attributes for a port, and could use
+  the selinux role.
+
+There are many more candidates which could use the Certificate role, the Network
+role, the Firewall role, etc.  Any role that needs an X509 cert could use the
+certificate role.  Any role that needs to open a port could use the firewall and
+selinux roles.
+
+# Tasks
+
+Roles will need to add tasks to enable/disable/add/remove the cert/port/whatever
+using the other role.  The role must specify the other roles using the FQRN e.g.
+```yaml
+# logging/tasks/main.yml
+  - name: Configure firewall
+    include_role:
+      name: linux-system-roles.firewall
+    vars:
+      firewall:
+        - port: 9876
+          state: enabled
+```
+
+# Dependencies/Requirements
+
+The role will need some way to specify the dependencies on the other roles.
+Unfortunately, we cannot use role dependencies in `meta/main.yml` as in
+(Role Dependencies)[https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#using-role-dependencies].
+The problem is that this not only *installs* the roles, but it also *executes*
+the roles. There is no way to turn off execution and just use dependency
+installation. In our roles we want to have the ability to call roles in-line
+and/or dynamically using the `roles` keyword and using `include_role` and using
+`import_role`. We need another mechanism, but there isn't another automatic
+dependency mechanism. So, just like we had to do for roles having dependencies
+on collections, and created our own `meta/collection-requirements.yml`, we
+will use a file `meta/role-requirements.yml`.  It will be in the same format as
+(requirements.yml)[https://docs.ansible.com/ansible/latest/galaxy/user_guide.html#installing-multiple-roles-from-a-file]
+For example:
+
+```yaml
+# logging/meta/role-requirements.yml
+roles:
+  - name: linux-system-roles.firewall
```

Comment:
I have two (and half :) questions...
1) Are only the linux system roles allowed here?
2) Could FQCN be put, as well? (Maybe, can't we just support FQCN and the collection format to make it simple???)

Reply:
> I have two (and half :) questions...
> 
>     1. Are only the linux system roles allowed here?

Yes and no.  The problem is that when we convert the role to collection, how do we convert `some-org.rolename` to the collection?  If we leave it as a role, I suppose we could have a collection depend on a role.

I would rather avoid this situation for now - we aren't using any roles other than system roles for now - if later we find that we have to use some external role, we can figure out what to do then.

> 
>     2. Could FQCN be put, as well? (Maybe, can't we just support FQCN and the collection format to make it simple???)

Yes.  We could use collection `include_role: name: fedora.linux_system_roles.ROLENAME` and then use `collection-requirements.yml` for the dependency - the only tricky part is converting to the downstream collection - when converting role ROLENAME we would have to know to look for  `include_role: name: fedora.linux_system_roles.OTHERROLE` and convert it to `include_role: name: redhat.rhel_system_roles.OTHERROLE`

Reply:
> (Maybe, can't we just support FQCN and the collection format to make it simple???)

Do you mean - let's not support legacy roles, and only support collections?  Works for me.  In fact, that's what Sergei is doing in https://github.com/linux-system-roles/mssql/pull/62/files#diff-448cbf29ac422b869521a75c0e601cb3e3d3e68ad4a9636ddd76551a6a2fddd4R29

Reply:
Yes! I was wondering if it causes any difficulties... If not, let's make it simple.

Reply:
> Yes! I was wondering if it causes any difficulties... If not, let's make it simple.

Do we include the `redhat.rhel_system_roles` collection in the RHEL7 RPM package?  If not, then we won't be able to `include_role: name: redhat.rhel_system_roles.firewall`.  Not sure what to do there.

One option would be for the RHEL7 RPM spec to do a `sed s/fedora.linux_system_roles/rhel-system-roles/g`

Reply:
But if you keep roles source in the legacy format and then convert them to the collection format in build, referencing a FQCN in the source is inconsistent. (And prevents you from solving donwstream naming by using `sed s/linux-system-roles/rhel-system-roles/` as the first step in the downstream build.)

Reply:
> But if you keep roles source in the legacy format and then convert them to the collection format in build, referencing a FQCN in the source is inconsistent.

Right, but I think that is a minor price to pay for the benefit of not having to invent and maintain a legacy role dependency mechanism.

> (And prevents you from solving donwstream naming by using `sed s/linux-system-roles/rhel-system-roles/` as the first step in the downstream build.)

The sed will be slightly different, but I don't think this will be a big problem.

Reply:
> Do we include the redhat.rhel_system_roles collection in the RHEL7 RPM package?

Yes, we do. RHEL7.9 spec file is almost identical to the latest released 8.x.